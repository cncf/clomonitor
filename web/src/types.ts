export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
