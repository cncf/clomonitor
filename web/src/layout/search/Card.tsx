import moment from 'moment';
import { GrPieChart } from 'react-icons/gr';
import { useNavigate } from 'react-router-dom';

import { Project } from '../../types';
import CartegoryBadge from '../common/badges/CategoryBadge';
import MaturityBadge from '../common/badges/MaturityBadge';
import ExternalLink from '../common/ExternalLink';
import Image from '../common/Image';
import RoundScore from '../common/RoundScore';
import ScoreSummary from '../common/ScoreSummary';
import styles from './Card.module.css';
import RepositorySection from './RepositorySection';

interface Props {
  project: Project;
  currentQueryString: string;
  saveScrollPosition: () => void;
}

const Card = (props: Props) => {
  const navigate = useNavigate();

  return (
    <div className={`col-12 col-sm-6 col-md-12 col-lg-6 col-xxxl-4 ${styles.cardWrapper}`} role="listitem">
      <div
        className={`card rounded-0 p-0 p-md-3 p-lg-0 p-xl-3 h-100 mw-100 d-flex text-reset text-decoration-none ${styles.card} card`}
        role="button"
        onClick={() => {
          props.saveScrollPosition();
          navigate(`/projects/${props.project.organization.name}/${props.project.name}`, {
            state: { currentSearch: props.currentQueryString },
          });
        }}
      >
        <div className="d-flex flex-column flex-sm-row align-items-center">
          <div
            className={`d-none d-md-flex d-lg-none d-xl-flex align-items-center justify-content-center ${styles.imageWrapper}`}
          >
            <Image alt={`${props.project.display_name || props.project.name} logo`} url={props.project.logo_url} />
          </div>
          <div className="flex-grow-1 ms-0 ms-md-3 ms-lg-0 ms-xl-3 w-100 truncateWrapper">
            <div className={`p-2 p-md-3 p-lg-2 p-xl-3 ${styles.content}`}>
              <div className="d-flex flex-row align-items-center">
                <div
                  className={`d-flex d-md-none d-lg-flex d-xl-none align-items-center justify-content-center me-3 ${styles.miniImageWrapper}`}
                >
                  <Image
                    alt={`${props.project.display_name || props.project.name} logo`}
                    url={props.project.logo_url}
                  />
                </div>
                <div className="d-flex flex-column w-100 truncateWrapper">
                  <div className="d-flex flex-row justify-content-between align-items-end">
                    <span className={`text-truncate fw-bold mb-0 ${styles.title}`}>
                      {props.project.display_name || props.project.name}
                    </span>
                  </div>

                  <div className="d-flex flex-row align-items-center my-2">
                    <MaturityBadge maturityLevel={props.project.maturity_id} />
                    <CartegoryBadge
                      categoryId={props.project.category_id}
                      className="d-none d-md-block d-lg-none d-xxl-block ms-2"
                    />
                  </div>

                  <div className={`d-none d-md-flex d-lg-none d-xl-flex flex-row align-items-center ${styles.info}`}>
                    <RepositorySection repositories={props.project.repositories} />

                    {props.project.devstats_url && (
                      <>
                        <ExternalLink label="Dev stats link" href={props.project.devstats_url} className="ms-3">
                          <div className={`d-flex flex-row align-items-center ${styles.link}`}>
                            <GrPieChart className={`me-1 ${styles.statsIcon}`} />
                            <div>DevStats</div>
                          </div>
                        </ExternalLink>
                      </>
                    )}
                  </div>
                </div>

                <div className="d-flex d-md-none d-lg-flex d-xl-none">
                  <RoundScore score={props.project.score.global} className={`ms-2 ${styles.global}`} />
                </div>
              </div>
            </div>
          </div>
        </div>

        <p className={`text-muted mx-3 my-3 my-md-4 ${styles.description}`}>{props.project.description}</p>

        <div className="mt-auto">
          <ScoreSummary score={props.project.score} bigSize={false} />
        </div>
        <div className={`d-none d-md-block d-lg-none d-xl-block text-end text-muted fst-italic mt-2 ${styles.legend}`}>
          Updated {moment.unix(props.project.updated_at).fromNow()}
        </div>
      </div>
    </div>
  );
};

export default Card;
