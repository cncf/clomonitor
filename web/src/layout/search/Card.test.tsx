import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import { RepositoryKind, ScoreKind } from '../../types';
import Card from './Card';

jest.mock('moment', () => ({
  ...(jest.requireActual('moment') as {}),
  unix: () => ({
    fromNow: () => '3 days ago',
  }),
}));

const mockSaveScrollPosition = jest.fn();

const defaultProps = {
  project: {
    category_id: 0,
    description:
      'Artifact Hub is a web-based application that enables finding, installing, and publishing packages and configurations for CNCF projects.',
    devstats_url: 'https://artifacthub.devstats.cncf.io/',
    display_name: 'Artifact Hub',
    id: '00000000-0001-0000-0000-000000000000',
    home_url: 'https://artifacthub.io',
    logo_url:
      'https://raw.githubusercontent.com/cncf/artwork/master/projects/artifacthub/icon/color/artifacthub-icon-color.svg',
    maturity_id: 2,
    name: 'artifact-hub',
    organization: {
      name: 'artifact-hub',
    },
    rating: 'a',
    repositories: [
      {
        kind: RepositoryKind.Primary,
        name: 'artifact-hub',
        url: 'https://github.com/artifacthub/hub',
      },
    ],
    score: {
      best_practices: 100,
      documentation: 75,
      global: 89,
      license: 80,
      legal: 75,
      score_kind: ScoreKind.Primary,
      security: 100,
    },
    updated_at: 1645138013,
  },
  currentQueryString: '',
  saveScrollPosition: mockSaveScrollPosition,
};

describe('Card', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <Router>
        <Card {...defaultProps} />
      </Router>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <Router>
          <Card {...defaultProps} />{' '}
        </Router>
      );

      const images = screen.getAllByAltText('Artifact Hub logo');
      expect(images).toHaveLength(2);
      expect(images[0]).toHaveProperty(
        'src',
        'https://raw.githubusercontent.com/cncf/artwork/master/projects/artifacthub/icon/color/artifacthub-icon-color.svg'
      );

      expect(screen.getByText('Artifact Hub')).toBeInTheDocument();
      expect(screen.getByTestId('maturity-badge')).toBeInTheDocument();
      expect(screen.getByText('Sandbox')).toBeInTheDocument();
      expect(screen.getByTestId('category-badge')).toBeInTheDocument();
      expect(screen.getByText('App definition')).toBeInTheDocument();

      const repoLink = screen.getByRole('link', { name: 'Repository link' });
      expect(repoLink).toHaveProperty('href', 'https://github.com/artifacthub/hub');

      const statsLink = screen.getByRole('link', { name: 'Dev stats link' });
      expect(statsLink).toHaveProperty('href', 'https://artifacthub.devstats.cncf.io/');

      const globalScores = screen.getAllByTestId('global-score');
      expect(globalScores).toHaveLength(2);
      expect(globalScores[1]).toHaveTextContent('89');
    });
  });
});
