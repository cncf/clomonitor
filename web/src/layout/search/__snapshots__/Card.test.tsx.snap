// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-12 col-sm-6 col-md-12 col-lg-6 col-xxxl-4 cardWrapper"
    role="listitem"
  >
    <div
      class="card rounded-0 p-0 p-md-3 p-lg-0 p-xl-3 h-100 mw-100 d-flex text-reset text-decoration-none card card"
      role="button"
    >
      <div
        class="d-flex flex-column flex-sm-row align-items-center"
      >
        <div
          class="d-none d-md-flex d-lg-none d-xl-flex align-items-center justify-content-center imageWrapper"
        >
          <img
            alt="Artifact Hub logo"
            aria-hidden="true"
            src="https://raw.githubusercontent.com/cncf/artwork/master/projects/artifacthub/icon/color/artifacthub-icon-color.svg"
          />
        </div>
        <div
          class="flex-grow-1 ms-0 ms-md-3 ms-lg-0 ms-xl-3 w-100 truncateWrapper"
        >
          <div
            class="p-2 p-md-3 p-lg-2 p-xl-3 content"
          >
            <div
              class="d-flex flex-row align-items-center"
            >
              <div
                class="d-flex d-md-none d-lg-flex d-xl-none align-items-center justify-content-center me-3 miniImageWrapper"
              >
                <img
                  alt="Artifact Hub logo"
                  aria-hidden="true"
                  src="https://raw.githubusercontent.com/cncf/artwork/master/projects/artifacthub/icon/color/artifacthub-icon-color.svg"
                />
              </div>
              <div
                class="d-flex flex-column w-100 truncateWrapper"
              >
                <div
                  class="d-flex flex-row justify-content-between align-items-end"
                >
                  <span
                    class="text-truncate fw-bold mb-0 title"
                  >
                    Artifact Hub
                  </span>
                </div>
                <div
                  class="d-flex flex-row align-items-center my-2"
                >
                  <div
                    class="badge text-secondary border border-secondary rounded-0 position-relative badge undefined"
                    data-testid="maturity-badge"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        class="me-2"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 512 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M439 32v165h18V32h-18zm-18 12.99L327.6 80l93.4 35V44.99zM165.9 103c-5 0-10.2 2.3-15.3 7-6.2 5.8-11.5 15.1-13.8 26.3-2.3 11.3-1 22 2.5 29.7 3.5 7.8 8.6 12.3 14.6 13.5 6 1.3 12.4-.9 18.7-6.6 6.1-5.8 11.5-15.1 13.8-26.4 2.2-11.3.9-22-2.5-29.7-3.5-7.8-8.6-12.2-14.6-13.5-1.1-.2-2.3-.3-3.4-.3zm-38.4 78.5c-3.4 1.2-6.9 2.5-10.7 4.1-24.85 15.7-42.2 31.2-59.84 55.7-11.19 15.5-11.74 42-12.58 61.5l20.8 9.2c.87-27.8.36-39.3 13.27-55.3 9.83-12.2 19.33-25 37.55-28.9 1.6 28.9-2.6 73.7-14 119.6 20.5 2.8 37.6-.7 57-6.3 50.7-25.3 74.1-3.8 109.3 45.7l20.5-32.1c-24.6-28.9-48.5-75.1-117.2-57.3 5-27.3 5.6-45.4 8.6-72.6.6-12 .8-23.9 1.1-35.7-8.9 6.8-19.9 10.4-31 8.1-9.5-2-17.3-7.9-22.8-15.7zm144.2 7.3c-18.2 17.8-22.2 31-50.2 38.4l-22.5-24c-.4 12.8-.8 25.9-1.9 39.2 9.5 8.7 19.2 15.7 22.7 14.6 31.3-9.4 40.3-20.3 61.4-41.9l-9.5-26.3zM409 215v96h-96v96h-96v78.1c102.3.2 167.8 1.1 270 1.8V215h-78zM140.7 363.9c-13.6 2.5-27.8 3.3-43.44.9-10.89 37.5-26.76 74.3-48.51 102.5l38.63 15.3c27.02-37.9 36.82-70.6 53.32-118.7z"
                        />
                      </svg>
                      Sandbox
                    </div>
                  </div>
                  <div
                    class="badge text-secondary border border-secondary rounded-0 position-relative badge d-none d-md-block d-lg-none d-xxl-block ms-2"
                    data-testid="category-badge"
                  >
                    <div
                      class="d-flex flex-row align-items-center"
                    >
                      <svg
                        class="me-2"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                        <path
                          d="M12 2l-5.5 9h11z"
                        />
                        <circle
                          cx="17.5"
                          cy="17.5"
                          r="4.5"
                        />
                        <path
                          d="M3 13.5h8v8H3z"
                        />
                      </svg>
                      App definition
                    </div>
                  </div>
                </div>
                <div
                  class="d-none d-md-flex d-lg-none d-xl-flex flex-row align-items-center info"
                >
                  <a
                    aria-label="Repository link"
                    class="link undefined"
                    href="https://github.com/artifacthub/hub"
                    rel="noopener noreferrer"
                    tabindex="-1"
                    target="_blank"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline"
                    >
                      <div
                        class="d-flex flex-row align-items-center link"
                      >
                        <svg
                          class="me-1 icon"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M12 0a12 12 0 1 0 0 24 12 12 0 0 0 0-24zm3.163 21.783h-.093a.513.513 0 0 1-.382-.14.513.513 0 0 1-.14-.372v-1.406c.006-.467.01-.94.01-1.416a3.693 3.693 0 0 0-.151-1.028 1.832 1.832 0 0 0-.542-.875 8.014 8.014 0 0 0 2.038-.471 4.051 4.051 0 0 0 1.466-.964c.407-.427.71-.943.885-1.506a6.77 6.77 0 0 0 .3-2.13 4.138 4.138 0 0 0-.26-1.476 3.892 3.892 0 0 0-.795-1.284 2.81 2.81 0 0 0 .162-.582c.033-.2.05-.402.05-.604 0-.26-.03-.52-.09-.773a5.309 5.309 0 0 0-.221-.763.293.293 0 0 0-.111-.02h-.11c-.23.002-.456.04-.674.111a5.34 5.34 0 0 0-.703.26 6.503 6.503 0 0 0-.661.343c-.215.127-.405.249-.573.362a9.578 9.578 0 0 0-5.143 0 13.507 13.507 0 0 0-.572-.362 6.022 6.022 0 0 0-.672-.342 4.516 4.516 0 0 0-.705-.261 2.203 2.203 0 0 0-.662-.111h-.11a.29.29 0 0 0-.11.02 5.844 5.844 0 0 0-.23.763c-.054.254-.08.513-.081.773 0 .202.017.404.051.604.033.199.086.394.16.582A3.888 3.888 0 0 0 5.702 10a4.142 4.142 0 0 0-.263 1.476 6.871 6.871 0 0 0 .292 2.12c.181.563.483 1.08.884 1.516.415.422.915.75 1.466.964.653.25 1.337.41 2.033.476a1.828 1.828 0 0 0-.452.633 2.99 2.99 0 0 0-.2.744 2.754 2.754 0 0 1-1.175.27 1.788 1.788 0 0 1-1.065-.3 2.904 2.904 0 0 1-.752-.824 3.1 3.1 0 0 0-.292-.382 2.693 2.693 0 0 0-.372-.343 1.841 1.841 0 0 0-.432-.24 1.2 1.2 0 0 0-.481-.101c-.04.001-.08.005-.12.01a.649.649 0 0 0-.162.02.408.408 0 0 0-.13.06.116.116 0 0 0-.06.1.33.33 0 0 0 .14.242c.093.074.17.131.232.171l.03.021c.133.103.261.214.382.333.112.098.213.209.3.33.09.119.168.246.231.381.073.134.15.288.231.463.188.474.522.875.954 1.145.453.243.961.364 1.476.351.174 0 .349-.01.522-.03.172-.028.343-.057.515-.091v1.743a.5.5 0 0 1-.533.521h-.062a10.286 10.286 0 1 1 6.324 0v.005z"
                          />
                        </svg>
                        <div>
                          Repository
                        </div>
                      </div>
                    </div>
                  </a>
                  <a
                    aria-label="Dev stats link"
                    class="link ms-3"
                    href="https://artifacthub.devstats.cncf.io/"
                    rel="noopener noreferrer"
                    tabindex="-1"
                    target="_blank"
                  >
                    <div
                      class="d-flex flex-row align-items-baseline"
                    >
                      <div
                        class="d-flex flex-row align-items-center link"
                      >
                        <svg
                          class="me-1 statsIcon"
                          fill="currentColor"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M10,23 C5.02943725,23 1,18.9705627 1,14 C1,9.02943725 5.02943725,5 10,5 L10,14 C10,14 10.7746374,14.7746374 11.1619561,15.1619561 L16.363961,20.363961 C14.7352814,21.9926407 12.4852814,23 10,23 Z M14,10 L14,1 C18.9705627,1 23,5.02943725 23,10 L14,10 Z M14,13 L22,13 C22,15.2037225 21.2079601,17.2224541 19.8930183,18.7870568 L14,13 Z"
                            fill="none"
                            stroke="#000"
                            stroke-width="2"
                          />
                        </svg>
                        <div>
                          DevStats
                        </div>
                      </div>
                    </div>
                  </a>
                </div>
              </div>
              <div
                class="d-flex d-md-none d-lg-flex d-xl-none"
              >
                <div
                  class="d-flex align-items-center justify-content-center rounded-pill value global ms-2 global"
                  data-testid="global-score"
                >
                  89
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <p
        class="text-muted mx-3 my-3 my-md-4 description"
      >
        Artifact Hub is a web-based application that enables finding, installing, and publishing packages and configurations for CNCF projects.
      </p>
      <div
        class="mt-auto"
      >
        <div
          class="align-items-center d-flex flex-column flex-md-row summary flex-lg-column flex-xl-row"
        >
          <div
            class="d-none d-md-block d-lg-none d-xl-block d-lg-none d-xl-block"
          >
            <div
              class="d-flex align-items-center justify-content-center rounded-pill value global me-0 me-sm-4 mb-2 mb-sm-0"
              data-testid="global-score"
            >
              89
            </div>
          </div>
          <div
            class="flex-grow-1 w-100 position-relative categories"
          >
            <div
              class="row"
            >
              <div
                class="col-6 text-truncate"
              >
                <div
                  class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
                >
                  <span
                    class="pe-1 d-inline-block position-relative icon"
                  >
                    <svg
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M11 5H6a2 2 0 00-2 2v11a2 2 0 002 2h11a2 2 0 002-2v-5m-1.414-9.414a2 2 0 112.828 2.828L11.828 15H9v-2.828l8.586-8.586z"
                        stroke-linecap="round"
                        stroke-linejoin="round"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <span
                    class="text-truncate d-none d-md-inline-block"
                  >
                    Documentation
                  </span>
                  <span
                    class="d-inline-block d-md-none text-truncate"
                  >
                    Docs
                  </span>
                </div>
                <div
                  class="d-flex flex-row bg-white position-relative border overflow-hidden line"
                >
                  <div
                    class="text-center fw-bold font-monospace value "
                  >
                    75
                  </div>
                  <div
                    class="flex-grow-1 w-100 position-relative"
                  >
                    <div
                      class="position-absolute top-0 start-0 bottom-0"
                      data-testid="line"
                      style="width: calc(75% - 5px);"
                    >
                      <div
                        class="position-absolute arrow "
                        data-testid="peak"
                      />
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="col-6 text-truncate"
              >
                <div
                  class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
                >
                  <span
                    class="pe-1 d-inline-block position-relative icon"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 14 16"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M7 4c-.83 0-1.5-.67-1.5-1.5S6.17 1 7 1s1.5.67 1.5 1.5S7.83 4 7 4zm7 6c0 1.11-.89 2-2 2h-1c-1.11 0-2-.89-2-2l2-4h-1c-.55 0-1-.45-1-1H8v8c.42 0 1 .45 1 1h1c.42 0 1 .45 1 1H3c0-.55.58-1 1-1h1c0-.55.58-1 1-1h.03L6 5H5c0 .55-.45 1-1 1H3l2 4c0 1.11-.89 2-2 2H2c-1.11 0-2-.89-2-2l2-4H1V5h3c0-.55.45-1 1-1h4c.55 0 1 .45 1 1h3v1h-1l2 4zM2.5 7L1 10h3L2.5 7zM13 10l-1.5-3-1.5 3h3z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                  <span
                    class="text-truncate"
                  >
                    License
                  </span>
                </div>
                <div
                  class="d-flex flex-row bg-white position-relative border overflow-hidden line"
                >
                  <div
                    class="text-center fw-bold font-monospace value "
                  >
                    80
                  </div>
                  <div
                    class="flex-grow-1 w-100 position-relative"
                  >
                    <div
                      class="position-absolute top-0 start-0 bottom-0"
                      data-testid="line"
                      style="width: calc(80% - 5px);"
                    >
                      <div
                        class="position-absolute arrow "
                        data-testid="peak"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="row mt-2"
            >
              <div
                class="col-6 text-truncate"
              >
                <div
                  class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
                >
                  <span
                    class="pe-1 d-inline-block position-relative icon"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 512 512"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 32c-70.7 0-128 57.3-128 128s57.3 128 128 128 128-57.3 128-128S326.7 32 256 32zm0 208c-44.2 0-80-35.8-80-80s35.8-80 80-80 80 35.8 80 80-35.8 80-80 80zM193.7 307.4c-19.1-8.1-36.2-19.6-50.8-34.3-1.4-1.4-2.8-2.8-4.1-4.3L64 400h96l48 80 48-105.8 25.5-56.2c-8.4 1.3-16.9 2-25.5 2-21.6 0-42.5-4.2-62.3-12.6zM373.3 268.9c-1.3 1.4-2.7 2.9-4.1 4.3-14.6 14.6-31.7 26.2-50.7 34.2L294 361.2l-21.9 48.4L304 480l48-80h96l-74.7-131.1z"
                      />
                    </svg>
                  </span>
                  <span
                    class="text-truncate"
                  >
                    Best Practices
                  </span>
                </div>
                <div
                  class="d-flex flex-row bg-white position-relative border overflow-hidden line"
                >
                  <div
                    class="text-center fw-bold font-monospace value "
                  >
                    100
                  </div>
                  <div
                    class="flex-grow-1 w-100 position-relative"
                  >
                    <div
                      class="position-absolute top-0 start-0 bottom-0"
                      data-testid="line"
                      style="width: 100%;"
                    />
                  </div>
                </div>
              </div>
              <div
                class="col-6 text-truncate"
              >
                <div
                  class="d-flex flex-row align-items-baseline text-muted fw-bold flex-nowrap title"
                >
                  <span
                    class="pe-1 d-inline-block position-relative icon"
                  >
                    <svg
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 2C9.243 2 7 4.243 7 7v2H6c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-1V7c0-2.757-2.243-5-5-5zM9 7c0-1.654 1.346-3 3-3s3 1.346 3 3v2H9V7zm9.002 13H13v-2.278c.595-.347 1-.985 1-1.722 0-1.103-.897-2-2-2s-2 .897-2 2c0 .736.405 1.375 1 1.722V20H6v-9h12l.002 9z"
                      />
                    </svg>
                  </span>
                  <span
                    class="text-truncate"
                  >
                    Security
                  </span>
                </div>
                <div
                  class="d-flex flex-row bg-white position-relative border overflow-hidden line"
                >
                  <div
                    class="text-center fw-bold font-monospace value "
                  >
                    100
                  </div>
                  <div
                    class="flex-grow-1 w-100 position-relative"
                  >
                    <div
                      class="position-absolute top-0 start-0 bottom-0"
                      data-testid="line"
                      style="width: 100%;"
                    />
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="d-none d-md-block d-lg-none d-xl-block text-end text-muted fst-italic mt-2 legend"
      >
        Updated 3 days ago
      </div>
    </div>
  </div>
</DocumentFragment>
`;
